### ISSUE TITLE
Annotations to describe requests

### ISSUE THREAD

**Issue by danilbykov** [softwaremill/tapir#805]
I would like to describe requests using annotations. For example,

```scala
@tapirrequest
final case class MyRequest(
  @query
  @description("field-description")
  field1: String,
  @query("customized-name")
  @deprecated
  field2: Int,
  @cookie("my-secret-token")
  secretToken: String
)
```
We can create `EndpointInput`s using annotations `@query`, `@header`, `@cookie`. Already defined annotations `@description` and `@deprecate` will help to fill `EndpointIO.Info`.

Annotations `@query`, `@header`, `@cookie` may have optional name parameter. If some field in case class is declared with annotation without explicit name then name of current field is used. This will help to reduce duplication when user needs to declare case class with names and types and then duplicate these names and types when declaring `EndpointInput`'s.

What do you think? I can probably contribute these changes.

**Comment by adamw** [comment]
I'm usually wary about annotations, but we can think about it :)

I'd have three main questions:
* how would the annotated form be converted into an `Endpoint` instance?
* what use-cases would be easier/more readable/succint comparing to the current way of specifying endpoints?
* what about responses and error responses? How would using an annotated request interact with these?

**Comment by danilbykov** [comment]
Let me start from your second question. Let's assume I have following case class which is request for some HTTP handler
```scala
final case class MyRequest1(
  field1: Int,
  field2: String,
  field3: Boolean
)
```
I think it is quite typical case when field names of case class are the same as query parameters in URL. So in this case I need to write following code to describe my request
```scala
val endpointInput = query[Int]("field1")
  .and(query[String]("field2"))
  .and(query[Boolean]("field3"))
  .mapTo(MyRequest1)
```
As you can see I wrote some code but this code is completely defined by structure of my case class.

If I add following annotation to my case class I can generate exactly the same code
```scala
@tapirrequest
final case class MyRequest1(
  @query
  field1: Int,
  @query
  field2: String,
  @query
  field3: Boolean
)
```
without duplication types and parameter names.

If it is needed then I can choose different names for field in class and parameter in URL
```scala
  @query("parameter-name")
  field1: Int
```

We can reuse already defined annotations `@description` and `@deprecated` for queries
```scala
  @query
  @description("field-description")
  @deprecated
  field1: Int
```
It would be equivalent to
```scala
  query[Int]("field1").description("field-description").deprecated
```

> how would the annotated form be converted into an Endpoint instance?

We need some macro which generates `EndpointInput` in companion object which then can be used in `Endpoint` definition.

> what about responses and error responses? How would using an annotated request interact with these?

Requests are a bit different here because it is quite common when request is built on couple headers and several query parameters. Responses are usually defined by just `jsonBody[SomeResponseClass]`. I don't see how it can be useful for responses.

**Comment by adamw** [comment]
Ok, a macro generating an `EndpointInput` could make sense.

Let's look at the implementations of `EndpointInput` and if they could be supported
* `query` and `header` are quite easy (as you described)
* what about path captures? these are often interleaved with constant paths, so would it be possible to map paths using annotations as well?
* bodies - to create an input describing the body, we need the target type, the codec format (media type), and the raw body type. I'm not sure if it would be possible to capture this using an annotation. Unless ... the annotations can have parameters which would describe the input? A fallback `@input(value = jsonBody[T]) val body: T` - would that work?)
* fixed header/method/path - these would probably be unsupported
* capturing all headers/query parameters - probably could work
* what about `auth`? these could probably be modifier annotations for other inputs, e.g. `@apiKeyAuth @query val  apiKey: String`. But then we could end up with invalid combinations such as `@apiKeyAuth val apiKey: String` without another input. Here we come to the fact that annotations lack any contextual type-safety :)

**Comment by danilbykov** [comment]
> what about path captures? these are often interleaved with constant paths, so would it be possible to map paths using annotations as well?

Yes, I also don't see how we can apply annotations to define path captures. You mentioned about interleaving with constant paths. Another not obvious situation is when someone rearranges fields with annotations `@path` in class. It looks like as isomorphic operation. But leads to rearranging path segments in URL and breaks HTTP route.

> bodies - to create an input describing the body, we need the target type, the codec format (media type), and the raw body type. I'm not sure if it would be possible to capture this using an annotation. Unless ... the annotations can have parameters which would describe the input?

We know the target type from case class so we need to know only the codec format and the raw body type. This annotation may look like
```scala
class body[R](val bodyType: RawBodyType[R], val cf: CodecFormat) extends StaticAnnotation
```
with aliases for common cases
```scala
class jsonbody() extends body(StringBody(UTF8), CodecFormat.Json())
class xmlbody() extends body(StringBody(UTF8), CodecFormat.Json())
```
Then in macro we can try to summon `Codec[R, T, CodecFormat]`. Don't know whether I will have issues with implementing this :)

> what about auth? these could probably be modifier annotations for other inputs, e.g. @apiKeyAuth @query val apiKey: String. But then we could end up with invalid combinations such as @apiKeyAuth val apiKey: String without another input. 

It looks like it is not hard to fix because we always can inspect another annotations for field in case class and return compilation error with clear error message like "Annotation `@apiKeyAuth` must be used only with annotations `@query`, `@header` and so on"

**Comment by adamw** [comment]
Yes, sounds doable :) Do you think the path restriction won't be a blocker? I mean, if you create a case class describing the whole request, wouldn't you want it to include anything that's captured from the path as well?

Maybe we could somehow allow the user to combine a path-input (e.g. `"x" / path[Int]() / "y" / path[String]()` with the rest. Maybe a class-level annotation somehow? But I don't have any great ideas here (yet ;) )

**Comment by danilbykov** [comment]
I submitted pull requests with annotations for each input type except paths.

> if you create a case class describing the whole request, wouldn't you want it to include anything that's captured from the path as well?

I would add something to capture path as well but we need to agree some solution. These annotations are just alternative way to describe requests and user still can use current approach without annotations. I just don't want to implement some bad solution which will be removed/replaced in future because it will also cause problems for users.

> Maybe we could somehow allow the user to combine a path-input (e.g. "x" / path[Int]() / "y" / path[String]() with the rest.

I thought about this. Let's assume we have following class
```scala
@tapirrequest
final case class Req(
  @path path1: Int,
  @query field: String,
  @path path2: String
)
```
For this class we can generate function
```scala
def (path1: Int, path2: String): EndpointInput[Req] = ??? // macro implementation
```
But I don't know how it can be combined with path definition like `"x" / path[Int]("path1") / "y" / path[String]("path2")`. It seems some kind of flatMap on `EndpointInput` is required for this.

> Maybe a class-level annotation somehow?

This can work probably. For example,
```scala
@tapirrequest("x/{path1}/y/{path2}")
final case class Req(
  @path path1: Int,
  @query field: String,
  @path path2: String
)
```
Should we use plain string in annotation `@tapirrequest` to describe path? Or some another type which already carries information about segments and dynamic variables? But even for plain string we can check during compilation that case class has fields with name `path1` and `path2` and these fields are annotated with `@path`.

**Comment by adamw** [comment]
Sorry again, I could have answered long ago. Too much happening at once :)

I guess the string approach is the best we can do here. But with strict compile-time checking:
* if there is a path given to `@tapirrequest`, each variable segment must have a corresponding `@path` field
* if there are `@path` fields, they all need to be in the path given to the top-level annotation

I'm also wondering about the top-level annotation name. What we are describing, is an endpoint input. And that is what we are deriving basing on the annotation. So maybe instead of `@tapirrequest`, we should have `@endpointInput`? And symmetrically, we could probably also have `@endpointOutput`, with a restricted set of annotations that can be used (and an additional one - status code). But maybe that's another PR.

**Comment by adamw** [comment]
Released in [0.17.0](https://github.com/softwaremill/tapir/releases/tag/v0.17.0)
